streams/client.go

```diff
@@ -90,15 +93,33 @@ func (client *client) publishEvents(events []publisher.Event) ([]publisher.Event
 		logp.Debug("kinesis", "sent %d records: %v", len(records), records)
 		observer.Dropped(dropped)
 		observer.Acked(len(okEvents))
+		if len(records) == 0 {
+			logp.Info("kinesis", "No records were mapped")
+			return nil, nil
+		}
 	}
 	logp.Debug("kinesis", "mapped to records: %v", records)
 	res, err := client.putKinesisRecords(records)
 	failed := collectFailedEvents(res, events)
-	if err != nil && len(failed) == 0 {
-		failed = events
+	if len(failed) == 0 {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case kinesis.ErrCodeLimitExceededException:
+			case kinesis.ErrCodeProvisionedThroughputExceededException:
+			case kinesis.ErrCodeInternalFailureException:
+				logp.Info("putKinesisRecords failed (api level, not per-record failure). Will retry all records.", err)
+				failed = events
+			default:
+				logp.Warn("putKinesisRecords persistent failure. Will not retry", err)
+			}
+		}
 	}
-	if len(failed) > 0 {
-		logp.Info("retrying %d events on error: %v", len(failed), err)
+	if err != nil || len(failed) > 0 {
+		dur := client.backoff.Duration()
```

Comment:
What I'm missing here how it increases over time, it looks like it's created from scratch

Reply:
This is the way the backoff library works. Each call to `Duration()` increases the backoff period.